File: legacy/client_flutter/lib/util/validation.dart
/// 가입/로그인 입력 유효성 — 서버 [`util/validation.js`](../../../../server/src/util/validation.js)
/// 와 동일 규칙. 사용자 즉시 피드백 + 가입 전 사전 차단용.
///
/// 정식 검증은 서버가 `join.json` 진입 시 한 번 더 수행 (방어적 중복).
library;

const int passwordMin = 4;
const int passwordMax = 12;
const int nickMin = 1;
const int nickMax = 20;

final RegExp _emailRe = RegExp(r'^[^\s@]+@[^\s@]+\.[^\s@]+$');

/// 반환값: null = 통과, 그 외 = 에러 코드 (`email_required`/`email_invalid`/`email_too_long`).
String? validateEmail(String? email) {
  if (email == null || email.isEmpty) return 'email_required';
  if (!_emailRe.hasMatch(email)) return 'email_invalid';
  if (email.length > 128) return 'email_too_long';
  return null;
}

/// 반환값: null = 통과, `password_required`/`password_length`.
String? validatePassword(String? pass) {
  if (pass == null || pass.isEmpty) return 'password_required';
  if (pass.length < passwordMin || pass.length > passwordMax) return 'password_length';
  return null;
}

/// 반환값: null = 통과, `nickname_required`/`nickname_length`.
String? validateNickname(String? nick) {
  if (nick == null) return 'nickname_required';
  final trimmed = nick.trim();
  if (trimmed.isEmpty) return 'nickname_required';
  if (trimmed.length < nickMin || trimmed.length > nickMax) return 'nickname_length';
  return null;
}

/// 반환값: null = 통과, `character_invalid`.
String? validateCharacter(int? ch) {
  if (ch == null || ch < 0 || ch > 2) return 'character_invalid';
  return null;
}

File: legacy/server/src/util/validation.js
// 가입/로그인 입력 유효성 — 기획 "번역할 한국어.txt" 기준
//   "아이디가 이메일형식이 아닙니다."
//   "비밀번호는 4~12자리입니다."
//   "별칭이 없거나 글자수가 너무 깁니다"

const EMAIL_RE = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

const PASSWORD_MIN = 4;
const PASSWORD_MAX = 12;
const NICK_MIN = 1;
const NICK_MAX = 20;

function validateEmail(email) {
  if (!email || typeof email !== 'string') return 'email_required';
  if (!EMAIL_RE.test(email)) return 'email_invalid';
  if (email.length > 128) return 'email_too_long';
  return null;
}

function validatePassword(pass) {
  if (!pass || typeof pass !== 'string') return 'password_required';
  if (pass.length < PASSWORD_MIN || pass.length > PASSWORD_MAX) return 'password_length';
  return null;
}

function validateNickname(nick) {
  if (!nick || typeof nick !== 'string') return 'nickname_required';
  const trimmed = nick.trim();
  if (trimmed.length < NICK_MIN || trimmed.length > NICK_MAX) return 'nickname_length';
  return null;
}

function validateCharacter(ch) {
  const n = Number(ch);
  if (!Number.isInteger(n) || n < 0 || n > 2) return 'character_invalid';
  return null;
}

module.exports = {
  validateEmail,
  validatePassword,
  validateNickname,
  validateCharacter,
  PASSWORD_MIN,
  PASSWORD_MAX,
  NICK_MIN,
  NICK_MAX,
};
